provider/awssd/aws_sd.go

```diff
@@ -530,20 +888,137 @@ func (p *AWSSDProvider) DeregisterInstance(service *sd.Service, ep *endpoint.End
 	for _, target := range ep.Targets {
 		log.Infof("De-registering an instance \"%s\" for service \"%s\" (%s)", target, *service.Name, *service.Id)
 
-		if !p.dryRun {
-			_, err := p.client.DeregisterInstance(&sd.DeregisterInstanceInput{
-				InstanceId: aws.String(p.targetToInstanceID(target)),
-				ServiceId:  service.Id,
-			})
+		targetName := target
+		if p.isAAndSRVService(service) {
+			prefSrv, _, err := p.srvDescrSplit(aws.StringValue(service.Description))
 			if err != nil {
 				return err
 			}
+			_, _, _, _, port, err := p.preferredSRVSplit(prefSrv)
+			if err != nil {
+				return err
+			}
+			targetName = targetName + ":" + port
+		}
+
+		err := p.DeregisterInstanceById(service.Id, aws.String(p.targetToInstanceID(targetName)))
+		if err != nil {
+			return err
 		}
 	}
 
 	return nil
 }
 
+// DeregisterInstanceById removes an instance from given service given the instance Id.
+func (p *AWSSDProvider) DeregisterInstanceById(serviceId *string, instanceId *string) error {
+	if !p.dryRun {
+		_, err := p.client.DeregisterInstance(&sd.DeregisterInstanceInput{
+			InstanceId: instanceId,
+			ServiceId:  serviceId,
+		})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RemoveServiceAndInstances deregisters all instances from the service and removes the service.
+func (p *AWSSDProvider) RemoveServiceAndInstances(svc *sd.Service) error {
+	instances, err := p.ListInstancesByServiceID(svc.Id)
+	if err != nil {
+		return err
+	}
+	for _, instance := range instances {
+		log.Infof("De-registering an instance \"%s\" for service \"%s\" (%s) due to wrong port", *instance.Id, *svc.Name, *svc.Id)
+		err = p.DeregisterInstanceById(svc.Id, instance.Id)
+		if err != nil {
+			return err
+		}
+	}
+	log.Infof("Deleting a service \"%s\" (%s)", *svc.Name, *svc.Id)
+	if !p.dryRun {
+		_, err := p.client.DeleteService(&sd.DeleteServiceInput{
+			Id: svc.Id,
+		})
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
This part fails in the real test. The issue is that the service cannot be deleted while instances are being removed.
```
ERRO[1843] ResourceInUse: Service contains registered instances; delete the instances before deleting the service
```

Reply:
Yes, I know that it could error. I left it there in order to avoid an additional if (and make the code more readeable). Eventually, the service will be deleted during the next iteration anyway. I added an if to avoid the error.